models/DataObject/ClassDefinition/Data/Relations/AbstractRelations.php

```diff
@@ -121,21 +121,41 @@ public function save(Localizedfield|AbstractData|\Pimcore\Model\DataObject\Objec
         }
 
         $data = $this->getDataFromObjectParam($object, $params);
+        $myCurrentRawRelations = $this->prepareMyCurrentRelations($object, $params);
+        $myNewRawRelations = [];
+        $ignoreClassId = null;
+        $classId = match (true) {
+        	$object instanceof Concrete => $object->getClassId(),
+        	$object instanceof AbstractData => $object->getObject()->getClassId(),
+        	$object instanceof Localizedfield => $object->getObject()->getClassId(),
+        	$object instanceof \Pimcore\Model\DataObject\Objectbrick\Data\AbstractData => $object->getObject()->getClassId(),
+        };
+        $db = Db::get();
+
+        if (null === $classId) {
+            throw new \Exception('Invalid object type');
+        }
+
         if ($data !== null) {
             $relations = $this->prepareDataForPersistence($data, $object, $params);
 
-            if (is_array($relations) && !empty($relations)) {
-                $db = Db::get();
-
+            /**
+             * FieldCollection don't support delta Updates of relations
+             * A FieldCollection Entry does not have a unique ID,
+             * so we need to delete all relations and insert them again
```

Comment:
That's unfortunate, i feel that the FC current approach itself could be a dead-lock factor 😿 

Leaving some notes, maybe somebody has some ideas.

In short, when swapping or removing collections/groups, the delta calculator get confused:
![Screenshot from 2024-05-08 18-48-56](https://github.com/pimcore/pimcore/assets/6014195/44abd7c3-e509-4cdd-a08d-88c97263a356)
since we check by collection per collection, if we move the collection with 1 item down (swap order 🔽 crossed in red ), it would calculate and determine that 4 rows with `position=0` are added and 1 is removed, and for the second collection that 1 item  is added and 4 removed, resulting in 10 queries (minus any same item in both collections), instead of total of 5 updates ( `position` 0 to 1 and viceversa).

So far, it always removed everything and re-saved, in this resulting in [2 "big" and deadlockish remove queries](https://github.com/pimcore/pimcore/blob/196d996ac40726fff3e6067fe9d66b4f816e5729/models/DataObject/Fieldcollection/Dao.php#L260-L261) plus 5 inserts, even when there are no relevant changes at all (as soon as it is minimally "dirty").

The thing is, the former attempt in this PR was working "almost fine" for removing,adding elements and moving the order(`index`) within the same collection(aka `position` column). 

An idea would be for FCs to create the same delta calculator but *reversed* and without strictly limit to position: 
1. loop in loop and unsets all the rows that are the same before and after save (unchanged rows).
2. A second loop in loop that check the existing ones without considering the position and set a part for potential update (NB: the ones with the same position are already out from the first check) and unset these for next loop.
3. A third loop, here the remaining are either small updates (eg.index change within the same position/group), addition or deletion.

Could it work or would it be a loop hell?

Reply:
- in the case of the example above, it would basically stop at second loop in loop.  (4 updates)
- In the case we add the same row (Le Mans) in the second group and swap, the Le Mans row would be stripped out on the first loop in loop (leave unchanged), the second loop would update the remaining 4 to change to position 0,  and third loop in loop, would add 1 Le Mans to collection in position 1.   (4 updates, 1 addition, 0 removal)
- In the case we delete 3 rows from collection 2 (pos:1) and add 2 to collection 1, the first and second loop in loop would do nothing and it would perform independently of the position (3 removals and 2 additions)
- In the case we delete 3 rows from collection 2 (pos:1) and add 2 to collection 1, with 1 same row in this operation, the first loop would do nothing, the second one would update by moving to the right position where it's added  (2 removals and 1 update and 1 addition)

Reply:
@dpfaffenbauer Beside this point, could you check the other inputs? Then it would LGTM and ready to be merged while i would try the approach proposed above to see if it works in a new branch/PR focusing the field collections
Thank you in advance!